modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java

```diff
@@ -1371,6 +1372,18 @@ protected void runQuery(GridCacheQueryInfo qryInfo) {
                 }
             }
             catch (Throwable e) {
+                if (X.hasCause(e, ClassNotFoundException.class) && !qry.keepBinary() && cctx.binaryMarshaller() &&
+                    !cctx.localNode().isClient() && !log.isQuiet()) {
+                    U.quietAndInfo(log, "Suggestion for the cause of ClassNotFoundException");
+                    U.quietAndInfo(log, "To disable, set -D" + IGNITE_QUIET + "=true");
+                    U.quietAndInfo(log, "  ^-- Ignite configured to use BinaryMarshaller but keepBinary is false for " +
+                        "request");
+                    U.quietAndInfo(log, "  ^-- Server node need to load definition of data classes. " +
+                        "It can be reason of ClassNotFoundException(consider IgniteCache.withKeepBinary to fix)");
+                    U.quietAndInfo(log, "Refer this page for detailed information: " +
+                        "https://apacheignite.readme.io/docs/binary-marshaller");
+                }
+
```

Comment:
I'm not sure if it *performance* suggestion.
Moreover performance suggestion logged on node start and this suggestion logged only for certain exception(ClassNotFoundException) with certain configuration(BinaryMarhsaller and keepBinary=false)